tls/tls_internal.h

```diff
@@ -173,6 +186,55 @@ typedef struct tls_handshake_t {
 	};
 } TlsHandshake;
 
+#if defined(DEBUG) && (DEBUG >= 1)
+
+/*
+ * Debug helper macro that checks if it's OK to use a particular union field.
+ * Current valid union field is tracked in a separate variable — tag.
+ *
+ * Initially, the tag value is expected to be zero. That means it's not decided
+ * yet what union field is used. First access to a protected union field with
+ * help of the macro, sets a tag to some non-zero value. Then access to any
+ * other union field fires a BUG_ON.
+ */
+#define TTLS_HS_generic_field(hs, field, tag)				\
+(*({									\
+	T_DBG3("%s:%d accessing hs->%s, tag: %d → %d\n",		\
+	       __func__, __LINE__, #field, (hs)->field##_tag, tag);	\
+	BUG_ON((hs)->field##_tag != 0 && (hs)->field##_tag != tag);	\
+	(hs)->field##_tag = tag;					\
+	&(hs)->field;							\
+}))
+
+/*
+ * Similar to TTLS_HS_generic_field, but without tag checking.
+ * Always resets the tag.
+ */
+#define TTLS_HS_generic_field_reset_tag(hs, field, tag)			\
+(*({									\
+	T_DBG3("%s:%d accessing hs->%s, tag %d → %d (override)\n",	\
+	       __func__, __LINE__, #field, (hs)->field##_tag, tag);	\
+	(hs)->field##_tag = tag;					\
+	&(hs)->field;							\
+}))
+
+#else
+
+#define TTLS_HS_generic_field(hs, fieldname, tag) ((hs)->fieldname)
+#define TTLS_HS_generic_field_reset_tag(hs, fieldname, tag) ((hs)->fieldname)
+
+#endif /* defined(DEBUG) && (DEBUG >= 1) */
+
+#define TTLS_HS_dhm_ctx(hs)	TTLS_HS_generic_field(hs, dhm_ctx, 1)
+#define TTLS_HS_ecdh_ctx(hs)	TTLS_HS_generic_field(hs, ecdh_ctx, 2)
+#define TTLS_HS_tmp_sha256(hs)	TTLS_HS_generic_field(hs, tmp_sha256, 3)
+
+#define TTLS_HS_dhm_ctx_reset_tag(hs) 					\
+	TTLS_HS_generic_field_reset_tag(hs, dhm_ctx, 1)
+
+#define TTLS_HS_ecdh_ctx_reset_tag(hs) 					\
+	TTLS_HS_generic_field_reset_tag(hs, ecdh_ctx, 2)
+
```

Comment:
Clever enough and I appreciate the debugging code. Do you think that we'll need the code later? I thought that this is single-time problem, which is unlikely to appear once to be fixed... I like the code, but I'm not sure whether it's useful. If you believe that it's good to have the debugging option, then just fix the issue with double ECDH point format.

Reply:
I'd like to have those macros in the code. They add a little noise to the code, but help to test whenever there are union field collisions. In release build macros are expanded to direct field accesses, so they are not slowing hot paths.

I like the macros because I came up with the idea. So I need an opinion other than mine to determine if they are useful. :-) I guess they may be useful if we'll need to pack fields into unions.

Reply:
There is old time agreement about Tempesta code is that we do our best to not to pollute the production code by code which is helpful for testing. We still have a lot of EXPORT statements just to satisfy out test module - that's not good, but also not so harmful. Debugging code is also a pollution, but we can't live without it.

However, with the ECDH point format discussion above, I though about a TLS unit test - why not to create an instance of necessary data structures in the unit test and just verify that all the members (at least on boudns of data structures) are safe? If somebody later change the members order - the unit test fails there there will be good comments what's wrong. What do you think?

Reply:
> a TLS unit test - why not to create an instance of necessary data structures in the unit test and just verify that all the members (at least on boudns of data structures) are safe?

Yeah, we should definitely check that. I've added a `BUILD_BUG_ON` statement testing that (in #1254).